File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/MarketplaceViewModel.swift
import Foundation
import Observation
import CryptoKit
import Flagship
import FlagshipAPI
import FlagshipCore

@Observable
@MainActor
public final class MarketplaceViewModel {
    public private(set) var state: LoadingState<[MarketplaceListing]> = .idle
    public var searchQuery: String = ""

    private let client: any ScreensClient
    /// Signing seam — production derives the OWNER IRK from the Keystore
    /// (biometric on every install). Tests inject a fixed key so the recorded
    /// signature is verifiable. Mirror of FrontPageViewModel.signer + Android
    /// MarketplaceViewModel.signer.
    private let signer: @MainActor (String) async throws -> Curve25519.Signing.PrivateKey
    private let now: () -> Int64

    public init(
        client: any ScreensClient,
        signer: (@MainActor (String) async throws -> Curve25519.Signing.PrivateKey)? = nil,
        now: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.client = client
        self.now = now
        self.signer = signer ?? { reason in try await Keystore.deriveIRK(reason: reason) }
    }

    public var filtered: [MarketplaceListing] {
        guard let listings = state.value else { return [] }
        if searchQuery.isEmpty { return listings }
        let q = searchQuery.lowercased()
        return listings.filter {
            $0.title.lowercased().contains(q)
                || $0.summary.lowercased().contains(q)
                || $0.creator.lowercased().contains(q)
        }
    }

    public func load() async {
        state = .loading
        do {
            let resp = try await client.marketplaceBrowse()
            state = .loaded(resp.listings)
        } catch {
            state = .failed(error.localizedDescription)
        }
    }

    // MARK: - Install (parity with Android MarketplaceViewModel + the webapp
    // installFromMarketplace flow)

    /// Per-install lifecycle for the detail screen's CTA.
    public enum InstallState: Equatable, Sendable {
        case idle
        case installing
        case succeeded(serviceId: String)
        /// The listing is paid and not yet owned (.com install gate → 402). The
        /// screen routes the owner to checkout for `priceUsdCents`.
        case paymentRequired(priceUsdCents: Int)
        case failed(String)
    }

    public private(set) var installState: InstallState = .idle
    public func resetInstall() { installState = .idle }

    /// Install `creator`/`slug` onto the box at `serverId` (the pod's FQDN).
    /// Two-step like Android / the webapp: fetch the full listing for its
    /// carried `manifestJson`, verify the manifest hashes to the listing's
    /// committed `manifestHash` (reject a swapped manifest), sign the canonical
    /// bytes with the OWNER IRK (biometric), and POST the envelope to the box.
    /// A paid app the account doesn't own surfaces as `.paymentRequired`.
    public func install(creator: String, slug: String, serverId: String) async {
        installState = .installing
        let detail: MarketplaceListingDetail
        do {
            detail = try await client.marketplaceFetchListing(creator: creator, slug: slug)
        } catch {
            installState = .failed(installErrorText(error, fallback: "Couldn't fetch the listing."))
            return
        }

        // The manifest is carried on the listing but not in the signed canonical
        // bytes; bind it by re-checking it hashes to the committed manifestHash.
        if !detail.manifestHash.isEmpty,
           detail.manifestHash.lowercased() != manifestSha256Hex(detail.manifestJson) {
            installState = .failed("Manifest hash mismatch — refusing to install a tampered listing.")
            return
        }

        let request = InstallServiceRequest(
            serverId: serverId,
            creator: creator,
            slug: slug,
            manifestJson: detail.manifestJson,
            addOwnerToMembership: true,
            issuedAt: now()
        )

        let key: Curve25519.Signing.PrivateKey
        do {
            key = try await signer("Install \(creator)/\(slug)")
        } catch {
            installState = .failed("Couldn't access your account key: \(error.localizedDescription)")
            return
        }

        let signature: Data
        do {
            signature = try key.signature(for: installServiceCanonicalBytes(request))
        } catch {
            installState = .failed("Couldn't sign the install order: \(error.localizedDescription)")
            return
        }

        do {
            let resp = try await client.installFromMarketplace(
                InstallServiceEnvelope(request: request, signature: HexUtil.encode(signature))
            )
            installState = .succeeded(serviceId: resp.serviceId)
        } catch let e as ScreensClientError {
            // Paid-app gate: .com returns 402 with a `price_usd_cents` body when
            // the app must be purchased first. Route the owner to checkout.
            if case .http(let status, let message) = e, status == 402 {
                installState = .paymentRequired(priceUsdCents: parsePriceUsdCents(message))
                return
            }
            installState = .failed(e.errorDescription ?? "That install didn't take. Try again in a moment.")
        } catch {
            installState = .failed(installErrorText(error, fallback: "Couldn't reach the box."))
        }
    }

    private func installErrorText(_ error: Error, fallback: String) -> String {
        if let e = error as? ScreensClientError { return e.errorDescription ?? fallback }
        return fallback
    }

    /// Best-effort parse of `price_usd_cents` out of a 402 body. Returns 0 when
    /// the field is absent/unparseable — the screen still shows a purchase CTA.
    private func parsePriceUsdCents(_ message: String) -> Int {
        guard let data = message.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let cents = obj["price_usd_cents"] as? Int else {
            return 0
        }
        return cents
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/MarketplaceViewModel.kt
// Marketplace browse + install. `load()` pulls the catalog from the pod's
// BFF; `install(...)` fetches the full listing (manifest lives only on the
// single-listing endpoint), builds an InstallServiceRequest, signs the
// canonical bytes with the OWNER IRK (biometric, never silent — same path as
// FrontPageViewModel), wraps it as an envelope, and POSTs it straight to the
// box's /api/services. Mirror of iOS MarketplaceViewModel + the
// MarketplaceDetailContainer install flow.

package com.flagshipserver.app.viewmodels

import com.flagshipserver.app.api.InstallServiceEnvelope
import com.flagshipserver.app.api.InstallServiceRequest
import com.flagshipserver.app.api.MarketplaceListing
import com.flagshipserver.app.api.ScreensClient
import com.flagshipserver.app.api.installServiceCanonicalBytes
import com.flagshipserver.app.api.manifestSha256Hex
import com.flagshipserver.app.core.HexUtil
import com.flagshipserver.app.keystore.Keystore
import com.google.crypto.tink.subtle.Ed25519Sign
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class MarketplaceViewModel(
    private val client: ScreensClient,
    /** Signing seam — production derives the OWNER IRK from the Keystore
     *  (biometric on every install). Tests inject a fixed keypair so the
     *  recorded signature is verifiable. Mirror of FrontPageViewModel.signer. */
    private val signer: suspend (reason: String) -> Ed25519Sign = { r -> Keystore.deriveIRK(r) },
    private val now: () -> Long = { System.currentTimeMillis() },
) {
    private val _state = MutableStateFlow<LoadingState<List<MarketplaceListing>>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<List<MarketplaceListing>>> = _state.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()
    fun setSearchQuery(q: String) { _searchQuery.value = q }

    /** Listings filtered by the current [searchQuery] (title / summary /
     *  creator, case-insensitive). Empty until [load] succeeds. */
    val filtered: List<MarketplaceListing>
        get() {
            val listings = (_state.value as? LoadingState.Loaded)?.value ?: return emptyList()
            val q = _searchQuery.value.trim().lowercase()
            if (q.isEmpty()) return listings
            return listings.filter {
                it.title.lowercase().contains(q) ||
                    it.summary.lowercase().contains(q) ||
                    it.creator.lowercase().contains(q)
            }
        }

    suspend fun load() {
        _state.value = LoadingState.Loading
        try {
            _state.value = LoadingState.Loaded(client.marketplaceBrowse().listings)
        } catch (t: Throwable) {
            _state.value = LoadingState.Failed(t.message ?: "couldn't load the marketplace")
        }
    }

    /** Per-install lifecycle for the detail screen's CTA. */
    sealed interface InstallState {
        data object Idle : InstallState
        data object Installing : InstallState
        data class Succeeded(val serviceId: String) : InstallState
        data class Failed(val message: String) : InstallState
    }

    private val _installState = MutableStateFlow<InstallState>(InstallState.Idle)
    val installState: StateFlow<InstallState> = _installState.asStateFlow()

    fun resetInstall() { _installState.value = InstallState.Idle }

    /**
     * Install [creator]/[slug] onto the box at [serverId] (the pod's FQDN).
     * Two-step like the webapp / iOS: fetch the full listing for its
     * `manifestJson`, then sign + POST. The signature is over the canonical
     * bytes — NOT the request JSON — so the daemon's recomputed bytes verify.
     */
    suspend fun install(creator: String, slug: String, serverId: String) {
        _installState.value = InstallState.Installing
        try {
            val detail = client.marketplaceFetchListing(creator, slug)
            // The manifest is carried on the listing but not in the signed
            // canonical bytes; bind it by re-checking it hashes to the listing's
            // committed manifest_hash before installing a byte of it.
            if (detail.manifestHash.isNotEmpty() &&
                !detail.manifestHash.equals(manifestSha256Hex(detail.manifestJson), ignoreCase = true)
            ) {
                _installState.value =
                    InstallState.Failed("manifest hash mismatch — refusing to install a tampered listing")
                return
            }
            val request = InstallServiceRequest(
                serverId = serverId,
                creator = creator,
                slug = slug,
                manifestJson = detail.manifestJson,
                addOwnerToMembership = true,
                issuedAt = now(),
            )
            val irk = signer("Install $creator/$slug")
            val signature = irk.sign(installServiceCanonicalBytes(request))
            val resp = client.installFromMarketplace(
                InstallServiceEnvelope(request = request, signature = HexUtil.encode(signature)),
            )
            _installState.value = InstallState.Succeeded(resp.serviceId)
        } catch (t: Throwable) {
            _installState.value = InstallState.Failed(t.message ?: "install failed")
        }
    }
}
